### ISSUE TITLE
Feature request: automate projector plugin labeling based on classifier feature columns

### ISSUE THREAD

**Issue by amarjeetsihota** [tensorflow/tensorboard#1065]
Hi 

Firstly congratulations on Tensorboard, it is fantastic tool (especially when demo-ing to executives).

We have been developing AI solutions for a four months now and have built two commercial DNN Classifiers (for stockbroking clients) 

**Tensorboard Projector and the DNN Classifier**
One of the wow moments  we get from clients is when we show the learnt embeddings for hiddenlayer_0 in tensorboard with labels.   We will perform a T-SNE in front of them and the clients can see the clusters of data that the classifier is using to make it’s decisions.  

In fact, the client will have a-ha moments when the see the inputs that are clustered together.  

I am not sure it is obvious to many people how powerful visualising the embeddings are even in a simple 2-3 hidden layer classifier.  

**Feature Request**
Our feature request is to automatically generate the tsv file for hiddenlayer0 for this class of classifer.

**Automating TSV file generation for HiddenLayer0**
Given most classifiers feature columns are one-hot category vectors, buckets and numeric columns, we have built an automatic tsv generator that 

1. Hooks into DNNClassifer using SessionRunHook
2. Grabs the dnn/input_from_feature_columns/input_layer/concat:0 tensor 
3. For each column type e.g. categorical, bucket, numeric, embedding
      a. Creates a label based on categorical_colunn definition

I have put the code we use on github (refer to https://github.com/amarjeetsihota/tensorboard.classifier.auto_tsv_file_maker/blob/master/README.md) for your reference,.

**Demo**
For demo purposes, I have built a simple classifier using the  @“Mental Health in Tech Survey” dataset in Kaggle, and saved the created a standalone tensorboard projector  github ( click <a href="https://rawgit.com/amarjeetsihota/agiledev-projector/master/index.html">here</a>  and choose the “Mental Health Classifer” from the tensor list

**Only a suggestion**
Feel free to delete this request if it is not in the general direction of the product, as this is  for your interest and I don't want to clog up the to-do list unnecessarily.

**DevSummit**
I attending the summit next week so if you are interested in what we are doing please feel free to get in contact and I will see you there!

email amarjeet@agiledevsolutions.com.au (Lead AI consultant and owner)

**Comment by nfelt** [comment]
Thanks for your message and for this suggestion, I'm glad you've been enjoying TensorBoard and have found it useful.  We're a bit short-staffed right now so I think we're unlikely to have time for building a feature like this ourselves.  Also, something specific to DNN hidden layers may not be quite general enough to include in TensorBoard proper.

However, I could see a general-purpose SessionRunHook that exports feature columns as projector-plugin compatible TSV files being a useful addition to the projector plugin python API (alongside, e.g. `visualize_embeddings()`).  If you (or anyone else) is interested in contributing a feature like that I think we'd be happy to review it.

**Comment by amarjeetsihota** [comment]
Hi Nick

 

I’d be happy to give it a go.

 

I didn’t intend my feature to be part of tensorflow as I agree it is specific to the estimators class.

 

I think it would be a good idea that the team who built the estimator classes produce a tsv file by default (or at least by a constructor parameter).  

 

At the tensorflow summit I had a discussion with both Justine Tunney and Mustapha Ispir to just mention it as an idea.

 

I think this would increase tensorboard usage in the small to medium enterprise market.

 

 

Amarjeet

**Comment by bileschi** [comment]
As TensorFlow is moving away from feature columns in favor of Keras & PreProcessing layers, this is unlikely to happen.

https://github.com/keras-team/governance/blob/master/rfcs/20190502-preprocessing-layers.md
